core/trie/proof.go

```diff
@@ -172,8 +194,272 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
 				return false
 			}
 			expectedHash = proofNode.Edge.Child
-			remainingPath.Truncate(proofNode.Edge.Path.Len())
+			remainingPath.Truncate(remainingPath.len - proofNode.Edge.Path.Len())
 		}
 	}
+
 	return expectedHash.Equal(value)
 }
+
+// VerifyRangeProof verifies the range proof for the given range of keys.
+// This is achieved by constructing a trie from the boundary proofs, and the supplied key-values.
+// If the root of the reconstructed trie matches the supplied root, then the verification passes.
+// If the trie is constructed incorrectly then the root will have an incorrect key(len,path), and value,
+// and therefore it's hash will be incorrect.
+// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
+func VerifyRangeProof(root *felt.Felt, keys, values []*felt.Felt, proofKeys, proofValues [2]*felt.Felt,
```

Comment:
Note, proof is 1 array, not 2. It got deduplicated into one.

Reply:
A single proof is an array, but we need two proofs, right?

Reply:
Well, the easiest way to generate the proof is with the left (start query actually, probably will come back later) and right key separately. Then it is combined into one set. I'm guessing this is done because a lot of the time some of the top nodes are going to be duplicated.

Reply:
Geth seems to store proofs nodes in a key-value store, so I guess starkware are thinking about dong something similar.

Reply:
Mind if I open a separate issue for this?

Reply:
https://github.com/NethermindEth/juno/issues/1898